docs/notebooks/code_overview.pct.py

```diff
@@ -0,0 +1,238 @@
+#
+# Acquisition functions that are suitable for greedily building batches of points can be defined using `GreedyAcquisitionFunctionBuilder` (or `SingleModelGreedyAcquisitionBuilder`) using a `prepare_acquisition_function` method that also accepts the points already chosen to be in the current batch.
+
+# %% [markdown]
+# ### `AcquisitionOptimizer`
+#
+# The `AcquisitionOptimizer` type definition represents an optimizer function that maximizes an acquisition function over a search space. Trieste provides a `generate_continuous_optimizer` function for generating gradient-based optimizers for continuous spaces and `automatic_optimizer_selector` for quickly selecting an appropriate optimizer.
+
+# %% [markdown]
+# ### `BayesianOptimizer` and `AskTellOptimizer`
+#
+# The `BayesianOptimizer` and `AskTellOptimizer` classes are the two Bayesian optimization interfaces provided by Trieste.
+#
+# `BayesianOptimizer` exposes an `optimize` method for running a Bayesian optimization loop with given initial datasets and models, and a given number of steps (and an optional early stop callback).
+#
+# `AskTellOptimizer` provides greater control over the loop, by providing separate `ask` and `tell` steps for suggesting query points and updating the models with new observations.
+
+# %% [markdown]
+# ## Extending the key types
+
+# %%
+from __future__ import annotations
+
+from typing import Optional
+
+import tensorflow as tf
+from trieste.types import TensorType
+
+# %% [markdown]
+# This section explains how to define new observers, model types and acqusition functions.
+
+# %% [markdown]
+# ### New observers
+#
+# Defining an observer with a single observation is as simple as defining a function that returns that observation:
+
+# %%
+from trieste.objectives.utils import mk_observer
+
+
+def simple_quadratic(x: TensorType) -> TensorType:
+    "A trivial quadratic function over :math:`[0, 1]^2`."
+    return -tf.math.reduce_sum(x, axis=-1, keepdims=True) ** 2
+
+
+observer = mk_observer(simple_quadratic)
+observer(tf.constant([[0, 1], [1, 1]], dtype=tf.float64))
+
+# %% [markdown]
+# A multi-observation observer can be constructed from multiple functions:
+
+# %%
+from trieste.objectives.utils import mk_multi_observer
+
+
+def simple_constraint(x: TensorType) -> TensorType:
+    "A trivial constraints over :math:`[0, 1]^2`."
+    return tf.math.reduce_min(x, axis=-1, keepdims=True)
+
+
+multiobserver = mk_multi_observer(
+    OBJECTIVE=simple_quadratic, CONSTRAINT=simple_constraint
+)
+multiobserver(tf.constant([[0, 1], [1, 1]], dtype=tf.float64))
+
+# %% [markdown]
+# ### New probabilistic model types
+#
+# Defining a new probabilistic model type simply involves writing a class that implements all the relevant methods (at the very least `predict` and `sample`). For clarity, it is best to also explicitly inherit from the supported feature protocols.
+
+# %%
+from trieste.models.interfaces import (
```

Comment:
trajectory samplers are advanced methods. Possibly doesn't matter, otherwise use for example a more trivial feature, eg `SupportGetObservationNoise?